pkg/platform/common/vmlayer/appinst.go

```diff
@@ -914,6 +917,33 @@ func (v *VMPlatform) UpdateAppInst(ctx context.Context, clusterInst *edgeproto.C
 	}
 }
 
+func (v *VMPlatform) ChangeAppInstDNS(ctx context.Context, app *edgeproto.App, appInst *edgeproto.AppInst, OldURI string, updateCallback edgeproto.CacheUpdateCallback) error {
+	// only VM-type deployment require any work
+	if app.Deployment != cloudcommon.DeploymentTypeVM {
+		return nil
+	}
+	// Add a new DNS entry for appInst
+	rootLBDetail, err := v.VMProvider.GetServerDetail(ctx, appInst.Uri)
+	if err != nil {
+		return fmt.Errorf("failed to get server detail for VMApp rootLB %s, %s", appInst.Uri, err)
+	}
+
+	updateCallback(edgeproto.UpdateTask, "Setting Up Load Balancer")
+	pp := edgeproto.TrustPolicy{}
+	err = v.SetupRootLB(ctx, appInst.Uri, appInst.Uri, &appInst.CloudletKey, &pp, rootLBDetail, appInst.EnableIpv6, updateCallback)
+	if err != nil {
+		return err
+	}
+
+	// Now delete old DNS entry
+	updateCallback(edgeproto.UpdateTask, "Deleting old DNS")
+	if err := v.VMProperties.CommonPf.DeleteDNSRecords(ctx, OldURI); err != nil {
+		log.SpanLog(ctx, log.DebugLevelInfra, "failed to delete DNS record", "uri", OldURI, "err", err)
+	}
```

Comment:
One thing that's going to be an issue when you add the GRPC API for CCRM is that the CCRM does not have the "old" state of the object available. That's only available here in the CRM because notify goes through a cache and can pull out the old state from the cache. In CCRM, we'd typically just send the new AppInst object as the GRPC Request object. Since you've already written the new state to etcd, CCRM also can only look up the new state from Etcd.

One way to deal with this is to make a custom object (instead of just AppInst) to pass via the GRPC call that can also carry the old URI. But that makes the API messy, it'd be better to keep the API as simple CRUD APIs between Controller and CCRM as much as possible.

Alternatively you could add an annotation to the AppInst which is the old URI. That may be useful information to have. It would also allow us to recover if CRM crashed before it was able to delete the old URI (because you've already written the new URI to etcd at that point, once the update process at the Controller times out the old URI will be lost forever). It's not really that big of a deal here if we fail to clean up the old DNS entry, of course, but it's worth considering.

A third option to actually do the DNS entry delete from the Controller once its finished waiting for the CRM to finish updating. I dislike this option though because it's the platform-specific code that's managing DNS entries, and the Controller doesn't really know whether the platform actually needs to delete a DNS entry or not. I think it's cleaner to keep that responsibility in the platform-specific code.

Reply:
This is a good point - I did notice CCRM apis just taking the object. Let me add annotations in order to have the record of old uri.